File: src/harmony/protocol/resumeEnvelope.ts
/** Shared resume watermark. Clients never decrement lastSeq. */

export type ResumeStatus = "ok" | "gap"

export type ResumeEnvelope = {
  type: "resume"
  status: ResumeStatus
  after: number
  min_after: number
}

export function resumeEnvelope(after: number, minAfter = 0): ResumeEnvelope {
  const safeAfter = Number.isFinite(after) ? Math.max(0, Math.trunc(after)) : 0
  const safeMin = Number.isFinite(minAfter) ? Math.max(0, Math.trunc(minAfter)) : 0
  if (safeAfter < safeMin) {
    return { type: "resume", status: "gap", after: safeAfter, min_after: safeMin }
  }
  return { type: "resume", status: "ok", after: safeAfter, min_after: safeMin }
}

export function parseResumeEnvelope(raw: unknown): ResumeEnvelope | null {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return null
  const record = raw as Record<string, unknown>
  const kind = record.type === "resume" || record.event === "resume"
  if (!kind) return null
  const status = record.status === "gap" ? "gap" : record.status === "ok" ? "ok" : null
  if (!status) return null
  const after = Number(record.after ?? 0)
  const minAfter = Number(record.min_after ?? 0)
  if (!Number.isFinite(after) || !Number.isFinite(minAfter)) return null
  return resumeEnvelope(after, minAfter)
}

/** Apply a newly observed seq. Duplicates and rewinds are ignored. */
export function applySeq(lastSeq: number, incoming: number): number {
  if (!Number.isFinite(incoming) || incoming <= 0) return lastSeq
  return incoming > lastSeq ? incoming : lastSeq
}

/** Gap jumps forward to min_after. Never goes backwards. */
export function nextAfter(lastSeq: number, envelope: ResumeEnvelope): number {
  if (envelope.status === "gap") return Math.max(lastSeq, envelope.min_after)
  return lastSeq
}

File: src/android/app/src/main/java/com/leoyuan/leophoneagent/relay/ResumeEnvelope.kt
package com.leoyuan.leophoneagent.relay

import org.json.JSONObject

data class ResumeEnvelope(
    val status: String,
    val after: Int,
    val minAfter: Int,
) {
    val isGap: Boolean get() = status == "gap"
}

object ResumeEnvelopes {
    fun of(after: Int, minAfter: Int): ResumeEnvelope {
        val safeAfter = after.coerceAtLeast(0)
        val safeMin = minAfter.coerceAtLeast(0)
        return if (safeAfter < safeMin) {
            ResumeEnvelope("gap", safeAfter, safeMin)
        } else {
            ResumeEnvelope("ok", safeAfter, safeMin)
        }
    }

    fun parse(data: String): ResumeEnvelope? {
        val obj = runCatching { JSONObject(data) }.getOrNull() ?: return null
        val kind = obj.optString("type") == "resume" || obj.optString("event") == "resume"
        if (!kind) return null
        val status = obj.optString("status")
        if (status != "ok" && status != "gap") return null
        return of(obj.optInt("after", 0), obj.optInt("min_after", 0))
    }

    fun nextAfter(lastSeq: Int, envelope: ResumeEnvelope): Int =
        if (envelope.isGap) maxOf(lastSeq, envelope.minAfter) else lastSeq

    fun toJson(envelope: ResumeEnvelope): JSONObject =
        JSONObject()
            .put("type", "resume")
            .put("event", "resume")
            .put("status", envelope.status)
            .put("after", envelope.after)
            .put("min_after", envelope.minAfter)
}
